### DISCUSSION TITLE
Big overhaul upcoming

### DISCUSSION THREAD

**Discussion by Bouni** [Bouni/kicad-jlcpcb-tools#74] [General]
Hi everyone,

I'm going to overhaul this plugin heavily to make it better and faster 😎 

My plans so far:

- Use an sqlite db per project instead of an csv. That should make lookups etc. fast and easy!
- Use threads where ever things take a while to execute. This will prevent the UI from lagging

Possible improvements:
- Make the parts db available as an API, that would prevent outages in case of JLC returning an 404

Let me know what pou think and what else you would love to see in the future!

**Comment by Bouni** [comment]
I started experimenting with icons and I'm very happy with the outcome so far 😎 

![image](https://user-images.githubusercontent.com/948965/141464271-46db84d7-7ccb-497b-bbb2-64752af03551.png)

The database for the part info storage is working, but needs some fine tuning.

My thoughts on which data has priority I came up with this:

- Reference, value, footprint, BOM and POS from the board always wins over data in the db
- If value or footprint changed in the board, the LCSC value in the db is cleared as it most likely no longer fits
- If an lcsc number is set on the board and on the database, the database wins
- The stock values will come from the API as they are always outdated in the library

**Comment by Bouni** [comment]
Here's a quick teaser what the new version will look like:

![image](https://user-images.githubusercontent.com/948965/141986428-ed7a129a-5bb5-4db7-99ff-917bcdc57f3a.png)

I make use of icons everywhere I can. Very statisfied so far 😄

**Comment by Bouni** [comment]
Almost done! 🥳 

Main window, not much changed here:
![image](https://user-images.githubusercontent.com/948965/143199632-00844e7c-f41e-439b-848f-e57e3e4ffc0b.png)

The new rotation correction manager:
![image](https://user-images.githubusercontent.com/948965/143199665-10940a49-2cc1-4e18-9521-a1cf52f71427.png)

Part selector dialog:
![image](https://user-images.githubusercontent.com/948965/143199741-2a9a145e-45fe-4df2-84c4-7eb3154355a2.png)

Part details dialog:
![image](https://user-images.githubusercontent.com/948965/143199766-6b60242c-d016-463d-8aca-c86f81084511.png)

I would really appreciate if some of you (@bvernoux @DanielO @arturo182😏 ) could test out the major-refactor branch!

Whats missing:
 - Cost calculation
 - Translations

**Comment by DanielO** [comment]
Looking good - I will try and test this in the next day or 2 (or maybe over the weekend).

One idea I had is a way to aggregate by package and part type, because those parts should (usually) be the same LSCS part.

For example if you have 10 0603 100n caps it would be nice if it were easy to select them at once so they can be edited together.

**Reply by Bouni** [reply]
With the new version you can naturaly sort all columns, so I guess if you sort by value, 100nF for example, you can Select all of them with Click - Shift - Click and assign a part to the selected group of parts.
Its unlikely that you have multiple packages for a single value, isn't it?
If so you still can Click - Ctrl - Click select your parts.

Do you see this a a valid solution?

**Reply by DanielO** [reply]
I think that would be reasonable, another possibility would be a "select alike parts" button perhaps - ie you click on a 100n 0805 and click the button and it selects all the parts which are the same. I think that would save a lot of effort & potential error.

**Reply by Bouni** [reply]
How strict should such a button work? Value and package must be equal to the first selection?

**Reply by DanielO** [reply]
I think so - that would cover the most common use case well.

**Comment by bvernoux** [comment]
I have tested it with KiCad 6 Nightly r25305 on Windows10 and all seems to work fine the startup and search is instant now so it is a very big improvement with MySQL database
The only issue found is the kicad-jlcpcb-tools\corrections\cpl_rotations_db.csv is not loaded/used anymore so we shall add specific stuff by hand with the "Rotation Manager"
- It will be great if we can import/use our own cpl_rotations_db.csv like before
- Also where those cpl_rotations_db.csv are stored (I suspect they are stored in the kicad-jlcpcb-tools\jlcpcb\parts.db to be confirmed) ?

**Reply by Bouni** [reply]
I totally forgot to read the local corrections into the database 🤦🏽‍♂️ 
Let me fix that today and try it again. I'm going to read the file into the database and afterwards rename the file to `cpl_rotations_db.csv.backup`
Any ou're right with your guess where they stored! theres a table called rotation that keeps all the corrections!

Maybe I should start to think about export import functionality to allow users to share their corrections!?

**Reply by Bouni** [reply]
@bvernoux auto import from CSV done! Open the Rotation Manager and the import is triggered

**Reply by bvernoux** [reply]
Yes it will be a must have to have a feature to export/import (csv file) to share/save users corrections
Also about stock it will be great to have a button to update it for visible items only (without doing a full refresh of the whole database)

**Comment by DanielO** [comment]
I gave it a quick spin and I found a few (fairly minor) things:
   * If you click any of the "Toggle" buttons it deselects the entry (which is annoying if you pressed it by mistake). It also jumps to the top of the list of parts (probably related I think)
   * The "Hide excluded .." buttons are not quite wide enough for the text on my system.
   * Clicking a "Hide excluded .." button makes it shrink in size vertically. The text still fits properly but it ends up smaller than the other buttons.

Here's a picture showing the shrunk button and truncate text: https://imgur.com/a/7t4QuRl

**Reply by Bouni** [reply]
Check my latest commit, should work as expected now

**Reply by bvernoux** [reply]
I'm testing latest version and the sort is not working on "JLCPCB Library" => "Stock" column (it was already the case before IIRC)
In fact it seems the sort on any column is not working correctly and not using natural sort like in the main windows(JLCPCB Tools) which works fine
This issue is here https://github.com/Bouni/kicad-jlcpcb-tools/issues/56 but I mention it here also as it is test on major refactoring
![image](https://user-images.githubusercontent.com/432709/143577834-39095815-443b-4bac-9c1f-edbb9eb28031.png)

**Reply by bvernoux** [reply]
Idea: It will be great to add a "Refresh Stock" button for main window(JLCPCB Tools) to refresh the stock of all items displayed

**Reply by Bouni** [reply]
@bvernoux I had the same thought but that woul result in a query for each part in the list and I fear that LCSC is not very amused if users start to bombard their API for Stock refresh

**Reply by bvernoux** [reply]
Yes I suspect sending tons of request on their API is not the best (except if we can merge them in one request if it is possible)
Could you explain how the stock is refreshed today ?
(I suspect it requires full download of database or maybe there's an other way?)

**Reply by Bouni** [reply]
I guess if nobody finds an API call that enables the fetch of multile parts at onece there's no way.
The current stock info comes from the database, so downloading it completely is the only option.

**Comment by Bouni** [comment]
Just in case someone wonders why no new commits are being pushed, I'm on vacation for one week, so expect nothing until the 6th or so 🙃
